pkg/traversal/traversal.go

```diff
@@ -0,0 +1,293 @@
+	return &traversalService{
+		logger: logger,
+		storer: storer,
+	}, nil
+}
+
+func (s *traversalService) TraverseChunkAddresses(
+	ctx context.Context,
+	reference swarm.Address,
+	encrypted bool,
+	chunkAddressFunc swarm.AddressIterFunc,
+) error {
+
+	isFile, e, metadata, err := s.checkIsFile(ctx, reference)
+	if err != nil {
+		return err
+	}
+
+	// reference address could be missrepresented as file when:
+	// - content size is 64 bytes (or 128 for encrypted reference)
+	// - second reference exists and is JSON (and not actually file metadata)
+
+	if isFile {
+
+		isManifest, m, err := s.checkIsManifest(ctx, reference, encrypted, e, metadata)
+		if err != nil {
+			return err
+		}
+
+		// reference address could be missrepresented as manifest when:
+		// - file content type is actually on of manifest type (manually set)
+		// - content was unmarshalled
+		//
+		// even though content could be unmarshaled in some case, iteration
+		// through addresses will not be possible
+
+		if isManifest {
+			// process as manifest
+
+			err = m.IterateAddresses(ctx, func(manifestNodeAddr swarm.Address) (stop bool) {
+				err := s.traverseChunkAddressesFromManifest(ctx, manifestNodeAddr, encrypted, chunkAddressFunc)
+				if err != nil {
+					stop = true
+				}
+				return
+			})
+			if err != nil {
+				return fmt.Errorf("traversal: iterate chunks: %s: %w", reference, err)
+			}
+
+			metadataReference := e.Metadata()
+
+			err = s.processBytes(ctx, metadataReference, chunkAddressFunc)
+			if err != nil {
+				return nil
+			}
+
+			if stop := chunkAddressFunc(reference); stop {
+				return nil
+			}
+
+		} else {
+			err = s.traverseChunkAddressesAsFile(ctx, reference, chunkAddressFunc, e)
+			if err != nil {
+				return err
+			}
+		}
+
+	} else {
+		err = s.processBytes(ctx, reference, chunkAddressFunc)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *traversalService) traverseChunkAddressesFromManifest(
+	ctx context.Context,
+	reference swarm.Address,
+	encrypted bool,
+	chunkAddressFunc swarm.AddressIterFunc,
+) error {
+
+	isFile, e, _, err := s.checkIsFile(ctx, reference)
+	if err != nil {
+		return err
+	}
+
+	if isFile {
+		err = s.traverseChunkAddressesAsFile(ctx, reference, chunkAddressFunc, e)
+		if err != nil {
+			return err
+		}
+	} else {
+		err = s.processBytes(ctx, reference, chunkAddressFunc)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *traversalService) traverseChunkAddressesAsFile(
+	ctx context.Context,
+	reference swarm.Address,
+	chunkAddressFunc swarm.AddressIterFunc,
+	e *entry.Entry,
+) (err error) {
+
+	bytesReference := e.Reference()
+
+	err = s.processBytes(ctx, bytesReference, chunkAddressFunc)
+	if err != nil {
+		// possible it was custom JSON bytes, which matches entry JSON
+		// but in fact is not file, and does not contain reference to
+		// existing address, which is why it was not found in storage
+		if !errors.Is(err, storage.ErrNotFound) {
+			return nil
```

Comment:
return err no??

Reply:
This was done in order to cover as much as possible edge-cases for requirements here.

As noted in use case issue (and discussed), one would have some address, whose chunks should be discovered, and unpinned. Since we have multiple types in Swarm, there is no exact way to know what type is actually (bytes, files and collections). For this reason, the code basically tries to "detect" what type of address it has, and then find all chunks for it.
This concrete part tries to cover case where someone used `/bytes` (or `/chunks`) API to upload JSON which is compatible with `pkg/collection/entry.Metadata`, and then used that address with another address to create new chunk, which unmarshals to `pkg/collection/entry.Entry`, but there reference used for file does not actually exists.
In that case, since the content in that reference address does not exist, for that side there are no chunks, so this is covers that.

I figure all of that might be far-fetched case, but if I am not mistaken with that, this should be fine.

Of course, another possibility is that somehow the bytes content for the file was removed, but not the metadata or root entry chunk (which could be possible), in which case maybe error should be returned?
Or maybe errors on other places should be checked for this same condition and ignored?